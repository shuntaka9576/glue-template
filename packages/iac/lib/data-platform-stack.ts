import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

type Props = {} & cdk.StackProps;

const GLUE_SCRIPT_LOCATION = 'glue/scripts';

export class DataPlatFormStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: Props) {
    super(scope, id, props);

    const ioBucket = new cdk.aws_s3.Bucket(this, 'IoBucket', {
      bucketName: `io-${this.region}-${this.account}`,
      blockPublicAccess: cdk.aws_s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const glueSystemBucket = new cdk.aws_s3.Bucket(this, 'guleBucket', {
      bucketName: `glue-system-${this.region}-${this.account}`,
      blockPublicAccess: cdk.aws_s3.BlockPublicAccess.BLOCK_ALL,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    new cdk.aws_s3_deployment.BucketDeployment(this, 'GlueScriptDeploy', {
      sources: [cdk.aws_s3_deployment.Source.asset('../src-glue/src')],
      destinationBucket: glueSystemBucket,
      destinationKeyPrefix: GLUE_SCRIPT_LOCATION,
    });

    const glueJobRole = new cdk.aws_iam.Role(this, 'GlueJobRole', {
      assumedBy: new cdk.aws_iam.CompositePrincipal(
        new cdk.aws_iam.ServicePrincipal('glue.amazonaws.com')
      ),
      managedPolicies: [
        cdk.aws_iam.ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AWSGlueServiceRole'
        ),
      ],
      inlinePolicies: {
        inlinePolicies: cdk.aws_iam.PolicyDocument.fromJson({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Action: ['s3:ListBucket'],
              Resource: [`arn:aws:s3:::${glueSystemBucket.bucketName}`],
            },
            {
              Effect: 'Allow',
              Action: 's3:*Object',
              Resource: [`arn:aws:s3:::${glueSystemBucket.bucketName}/*`],
            },
            {
              Effect: 'Allow',
              Action: ['s3:ListBucket'],
              Resource: [`arn:aws:s3:::${ioBucket.bucketName}`],
            },
            {
              Effect: 'Allow',
              Action: 's3:*Object',
              Resource: [`arn:aws:s3:::${ioBucket.bucketName}/*`],
            },
          ],
        }),
      },
    });

    const jobNames = ['convert-fruits-job'];

    jobNames.map((jobName) => {
      const script_name = jobName.replace(/-/g, '_');

      new cdk.aws_glue.CfnJob(this, `${jobName}`, {
        name: `${jobName}`,
        command: {
          name: 'glueetl',
          pythonVersion: '3',
          scriptLocation: `s3://${glueSystemBucket.bucketName}/${GLUE_SCRIPT_LOCATION}/${script_name}.py`,
        },
        role: glueJobRole.roleArn,
        numberOfWorkers: 2,
        glueVersion: '4.0',
        workerType: 'G.1X',
        defaultArguments: {
          '--enable-glue-datacatalog': 'TRUE',
          '--enable-metrics': 'TRUE',
          '--enable-spark-ui': 'TRUE',
          '--enable-job-insights': 'FALSE',
          '--enable-continuous-cloudwatch-log': 'TRUE',
          '--job-language': 'python',
          '--job-bookmark-option': 'job-bookmark-disable',
          '--TempDir': `s3://${glueSystemBucket.bucketName}/glue/temporary/`,
          '--spark-event-logs-path': `s3://${glueSystemBucket.bucketName}/glue/sparkHistoryLogs/`,
          '--IO_S3_BUCKET': ioBucket.bucketName,
        },
      });
    });
  }
}
