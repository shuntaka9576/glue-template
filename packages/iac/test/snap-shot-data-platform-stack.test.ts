import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DataPlatFormStack } from '../lib/data-platform-stack';

test('snapshot test', () => {
  const app = new cdk.App();
  const stack = new DataPlatFormStack(app, 'data-platform-stack');
  const template = Template.fromStack(stack).toJSON();

  template.Resources.GlueScriptDeployCustomResourceFD1DDCB8.Properties.SourceObjectKeys =
    null;

  expect(template).toMatchSnapshot();
});
