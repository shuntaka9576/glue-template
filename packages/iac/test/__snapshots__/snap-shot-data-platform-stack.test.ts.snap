// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e976a796f036a5efbf44b99e44cfb5a961df08d8dbf7cd37e60bf216fb982a00.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_CA_BUNDLE": "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "GlueScriptDeployAwsCliLayer3723E96B",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "guleBucket6A83BDCB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "guleBucket6A83BDCB",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b7f33614a69548d6bafe224d751a7ef238cde19097415e553fe8b63a4c8fd8a6.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "IoBucket176E52CE",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GlueJobRoleF1B69418": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSGlueServiceRole",
              ],
            ],
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "s3:ListBucket",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:s3:::",
                        {
                          "Ref": "guleBucket6A83BDCB",
                        },
                      ],
                    ],
                  },
                },
                {
                  "Action": "s3:*Object",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:s3:::",
                        {
                          "Ref": "guleBucket6A83BDCB",
                        },
                        "/*",
                      ],
                    ],
                  },
                },
                {
                  "Action": "s3:ListBucket",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:s3:::",
                        {
                          "Ref": "IoBucket176E52CE",
                        },
                      ],
                    ],
                  },
                },
                {
                  "Action": "s3:*Object",
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:s3:::",
                        {
                          "Ref": "IoBucket176E52CE",
                        },
                        "/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "inlinePolicies",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GlueScriptDeployAwsCliLayer3723E96B": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "3322b7049fb0ed2b7cbb644a2ada8d1116ff80c32dca89e6ada846b5de26f961.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "GlueScriptDeployCustomResourceFD1DDCB8": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketKeyPrefix": "glue/scripts",
        "DestinationBucketName": {
          "Ref": "guleBucket6A83BDCB",
        },
        "Prune": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        ],
        "SourceObjectKeys": null,
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "IoBucket176E52CE": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketName": {
          "Fn::Join": [
            "",
            [
              "io-",
              {
                "Ref": "AWS::Region",
              },
              "-",
              {
                "Ref": "AWS::AccountId",
              },
            ],
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "IoBucketAutoDeleteObjectsCustomResource931F0B37": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "IoBucketPolicy1F10DB00",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "IoBucket176E52CE",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "IoBucketPolicy1F10DB00": {
      "Properties": {
        "Bucket": {
          "Ref": "IoBucket176E52CE",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "IoBucket176E52CE",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "IoBucket176E52CE",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "convertfruitsjob": {
      "Properties": {
        "Command": {
          "Name": "glueetl",
          "PythonVersion": "3",
          "ScriptLocation": {
            "Fn::Join": [
              "",
              [
                "s3://",
                {
                  "Ref": "guleBucket6A83BDCB",
                },
                "/glue/scripts/convert_fruits_job.py",
              ],
            ],
          },
        },
        "DefaultArguments": {
          "--IO_S3_BUCKET": {
            "Ref": "IoBucket176E52CE",
          },
          "--TempDir": {
            "Fn::Join": [
              "",
              [
                "s3://",
                {
                  "Ref": "guleBucket6A83BDCB",
                },
                "/glue/temporary/",
              ],
            ],
          },
          "--enable-continuous-cloudwatch-log": "TRUE",
          "--enable-glue-datacatalog": "TRUE",
          "--enable-job-insights": "FALSE",
          "--enable-metrics": "TRUE",
          "--enable-spark-ui": "TRUE",
          "--job-bookmark-option": "job-bookmark-disable",
          "--job-language": "python",
          "--spark-event-logs-path": {
            "Fn::Join": [
              "",
              [
                "s3://",
                {
                  "Ref": "guleBucket6A83BDCB",
                },
                "/glue/sparkHistoryLogs/",
              ],
            ],
          },
        },
        "GlueVersion": "4.0",
        "Name": "convert-fruits-job",
        "NumberOfWorkers": 2,
        "Role": {
          "Fn::GetAtt": [
            "GlueJobRoleF1B69418",
            "Arn",
          ],
        },
        "WorkerType": "G.1X",
      },
      "Type": "AWS::Glue::Job",
    },
    "guleBucket6A83BDCB": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketName": {
          "Fn::Join": [
            "",
            [
              "glue-system-",
              {
                "Ref": "AWS::Region",
              },
              "-",
              {
                "Ref": "AWS::AccountId",
              },
            ],
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
          {
            "Key": "aws-cdk:cr-owned:glue/scripts:acd539e9",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "guleBucketAutoDeleteObjectsCustomResource2CF51E62": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "guleBucketPolicy38A97307",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "guleBucket6A83BDCB",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "guleBucketPolicy38A97307": {
      "Properties": {
        "Bucket": {
          "Ref": "guleBucket6A83BDCB",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "guleBucket6A83BDCB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "guleBucket6A83BDCB",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
